File: 2024/YounSangHo/Week7/2024-11-10-implement-prefix-tree.js
class TrieNode {
  constructor() {
    this.children = new Map();
    this.endOfword = false;
  }
}

class PrefixTree {
  constructor() {
    this.root = new TrieNode();
  }

  /**
   * @param {string} word
   * @return {void}
   */
  insert(word) {
    let cur = this.root;
    for (let c of word) {
      if (!cur.children.has(c)) {
        cur.children.set(c, new TrieNode());
      }
      cur = cur.children.get(c);
    }

    cur.endOfword = true;
  }

  /**
   * @param {string} word
   * @return {boolean}
   */
  search(word) {
    let cur = this.root;
    for (let c of word) {
      if (!cur.children.has(c)) {
        return false;
      }
      cur = cur.children.get(c);
    }
    return cur.endOfword;
  }

  /**
   * @param {string} prefix
   * @return {boolean}
   */
  startsWith(prefix) {
    let cur = this.root;
    for (let c of prefix) {
      if (!cur.children.has(c)) {
        return false;
      }
      cur = cur.children.get(c);
    }

    return true;
  }
}

File: 2024/Jooyoung/2024-11-10-ImplementPrefixTree.py
class PrefixNode:
    def __init__(self) -> None:
        self.children = {}
        self.end_node = False


class PrefixTree:
    def __init__(self):
        self.root = PrefixNode()

    def insert(self, word: str) -> None:
        cur = self.root

        for w in word:
            if w not in cur.children:
                cur.children[w] = PrefixNode()

            cur = cur.children[w]

        cur.end_node = True

    def search(self, word: str) -> bool:
        cur = self.root

        for w in word:
            if w not in cur.children:
                return False

            cur = cur.children[w]

        return cur.end_node

    def startsWith(self, prefix: str) -> bool:
        cur = self.root

        for w in prefix:
            if w not in cur.children:
                return False

            cur = cur.children[w]

        return True

prefixTree = PrefixTree()
prefixTree.insert("dog")
prefixTree.search("dog")
prefixTree.search("do")
prefixTree.startsWith("do")
prefixTree.insert("do")
prefixTree.search("do")
